### PR TITLE
Notification module, Twilio integration and PHP unit tests for the module

### PR DESCRIPTION
**Pull request by fatimarafiqui**
This includes the notification module for Circle of Trust and Unit tests for the same module.
notifica

notification

### CODE REVIEW

**Comment by fatimarafiqui** [comment]
@prathmeshranaut @MehaKaushik Please review this. Thanks!

File: modules/Notification.php [hunk 1]
```diff
@@ -1,10 +1,98 @@
...
+            $r['message'] = 'Please provide a valid message.';
+        } else {
+            $client = new Client($this->account_sid, $this->auth_token);
+
+            $twilioResponse = $client->messages->create(
+                $number,
+                array(
+                    // A Twilio phone number you purchased at twilio.com/console
+                    'from' => $this->from_number,
+                    // the body of the text message you'd like to send
```
Comment (MehaKaushik) [inline modules/Notification.php]: Not a good style of writing comments. Do not use second person reference. Person reading your code is different from person using using your app. 
A more suggestive comment :
"The body of the text message, the logged in user would send to other comrades."

File: request/send_notification.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                    if (!empty($_POST['msg_type']) &&
+                        in_array($_POST['msg_type'], $types)
+                    ) {
+                        $msg_type = $_POST['msg_type'];
+                        $message = '';
+                        switch ($msg_type) {
+                            case $types[0]:
+                                $message = "Come and get me.I need help getting home safely.
+                                Call ASAP to get my Location.
+                                via FirstAide web app";
```
Comment (MehaKaushik) [inline request/send_notification.php]: Were these sample messages provided in requirements or mocks ?

File: tests/modules/NotificationTest.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    public function testSendSms()
+    {
+        $notification = new Notification(Notification::USE_TEST_CREDENTIALS);
+        $sendNotificationRespose = $notification->sendSms('+910000000000', 'Test message');
+        $this->assertTrue($sendNotificationRespose['response']);
+    }
+
+    /**
+     * Method : testGetFullPage
+     * Description : tests for the full card template content page
```
Comment (MehaKaushik) [inline tests/modules/NotificationTest.php]: Are the comments correct here ?

File: tests/modules/NotificationTest.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+$APPLICATION_DIR = str_replace('tests/modules', '', dirname(__FILE__));
+require_once $APPLICATION_DIR.'/includes/settings.php';
+
+/**
+ * Importing required modules/wrappers
+ */
+use FirstAide\Notification;
+use PHPUnit\Framework\TestCase;
+
+class NotificationTest extends TestCase
```
Comment (MehaKaushik) [inline tests/modules/NotificationTest.php:17]: Are these complete set of tests, which you plan to write for this module? or are there any more coming up ?

File: request/send_notification.php (hunk 2 above)
Comment (fatimarafiqui) [reply request/send_notification.php]: @MehaKaushik They were provided in the requirements, and Akanksha used them last year too.

File: tests/modules/NotificationTest.php (hunk 4 above)
Comment (fatimarafiqui) [reply tests/modules/NotificationTest.php:17]: @MehaKaushik According to me, these tests cover all the methods of the notification module. Therefore, their coverage is 100%. Do you have any other test suggestion?

File: modules/User.php [hunk 5]
```diff
@@ -171,6 +171,22 @@ public function getCircleOfTrust()
         }
         return false;
     }
+
+    public function getCircleOfTrustNumbers()
```
Comment (npari) [inline modules/User.php]: Lets add block comments to all methods

File: request/send_notification.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
+<?php
```
Comment (npari) [inline request/send_notification.php:1]: Lets make this file modular. Very long switch case.

Comment (npari) [reply request/send_notification.php:1]: Comments need to be added by default for all PRs.

File: tests/modules/NotificationTest.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+{
+    /**
+     * Method : testSendSms
+     * Description : test to check of the response for sending a sms is a success
+     */
+    public function testSendSms()
```
Comment (npari) [inline tests/modules/NotificationTest.php:63]: Can we add negative tests as well?

Comment (npari) [reply tests/modules/NotificationTest.php:63]: Add asserts when invalid message/number is passed in sendSms.

File: tests/modules/NotificationTest.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        $notification = new Notification(Notification::USE_TEST_CREDENTIALS);
+        $sendNotificationRespose = $notification->sendSms('+910000000000', 'Test message');
+        $this->assertTrue($sendNotificationRespose['response']);
```
Comment (npari) [inline tests/modules/NotificationTest.php:67]: In addition to checking if there is a response, asserts needs to be added to check if the expected response is received in positive and negative case.

File: tests/modules/NotificationTest.php [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * Method : testSendMultipleSms
+     * Description : tests for checking sms sent to multiple numbers
+     */
+    public function testSendMultipleSms()
+    {
+        $arrayOfTestNumbers = array(
```
Comment (npari) [inline tests/modules/NotificationTest.php:76]: Same comments as above

**Review by npari — CHANGES_REQUESTED** [review]

File: tests/modules/NotificationTest.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
+<?php
+/**
+ *
+ * NotificationTest.php is unit test file for Notification module (modules/Notification.php).
```
Comment (npari) [inline tests/modules/NotificationTest.php:4]: You can add a test to see if the account_sid, auth_token and from_number are getting set in the constructor as expected

File: tests/modules/NotificationTest.php (hunk 4 above)
Comment (npari) [reply tests/modules/NotificationTest.php:17]: 100% test coverage indicates every branch is tested in a If/else block, every line of code is tested, and every method is tested.

File: modules/User.php (hunk 5 above)
Comment (npari) [reply modules/User.php]: Remove commented pieces of code if irrelevant
```
 // Returns country from the given list of countries only
            /*$country_found = in_array($country_found, $active_post_countries) ?
                $country_found :
                'UG';
            */
```

File: tests/modules/NotificationTest.php (hunk 10 above)
Comment (fatimarafiqui) [reply tests/modules/NotificationTest.php:4]: @npari Twilio is handling this for us. This is not required from our end.

File: tests/modules/NotificationTest.php (hunk 8 above)
Comment (fatimarafiqui) [reply tests/modules/NotificationTest.php:67]: Done. https://github.com/systers/FirstAide-web/pull/254/commits/a7303aabfeb81f477cff586a72a0d3da7a71cc85

File: tests/modules/NotificationTest.php (hunk 7 above)
Comment (fatimarafiqui) [reply tests/modules/NotificationTest.php:63]: Done https://github.com/systers/FirstAide-web/pull/254/commits/a7303aabfeb81f477cff586a72a0d3da7a71cc85

File: tests/modules/NotificationTest.php (hunk 4 above)
Comment (fatimarafiqui) [inline tests/modules/NotificationTest.php]: I have added all possible tests.

Comment (fatimarafiqui) [reply tests/modules/NotificationTest.php:17]: Added more tests. @npari @MehaKaushik

File: request/send_notification.php (hunk 6 above)
Comment (fatimarafiqui) [reply request/send_notification.php:1]: Done here - https://github.com/systers/FirstAide-web/pull/254/commits/1bc0c1be80f529decd2b1ec00c6be7599c466058

File: modules/User.php (hunk 5 above)
Comment (fatimarafiqui) [reply modules/User.php]: Done https://github.com/systers/FirstAide-web/pull/254/commits/ef9beed758e1fa026c3fccbadb3e2829e3e7e9d4

File: tests/modules/NotificationTest.php (hunk 9 above)
Comment (fatimarafiqui) [reply tests/modules/NotificationTest.php:76]: Done.

**Review by npari — APPROVED** [review]

**Pull request merged**